### PR TITLE
Create metrics to track Scheduled->Queued->Running task state transition times

### PR DESCRIPTION
**Pull request by vandonr-amz**
The goal of this PR is to provide metrics that are more insightful than the often-misunderstood and very wide "task landing time"

I made a little graph to illustrate the different timings we currently have access to:
image
the landing time in purple is only available in the airflow UI. In black are the metrics currently being sent, and I'm proposing to add the orange one in this PR (actually it's not adoption time and it'll be 2 metrics but I don't want to re-do the graph).

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -1330,6 +1330,18 @@ def check_and_change_state_before_execution(
 
         if not test_mode:
             session.add(Log(State.RUNNING, self))
+
+        if not self.end_date:
+            # if the task has an end date, it means that this is not its first round.
+            # we send the adoption time metric only on the first try, otherwise it gets more complex.
+            adoption_time = (timezone.utcnow() - self.start_date).total_seconds()
+            Stats.timing(f"dag.{self.task.dag_id}.{self.task.task_id}.adoption_time", adoption_time)
```
Comment (ferruzzi) [inline airflow/models/taskinstance.py]: Any reason not to include tags on this one?

**Comment by ferruzzi** [comment]
I want to suggest finding a home for that diagram in the docs, I really like that.

**Comment by jedcunningham** [comment]
I think adding this metric is a good idea. However, that's not "adoption" time. Adoption refers to tasks a scheduler adopts from a no-longer-running scheduler.

It also may make sense to be more granular here and tract scheduled -> queued and queued -> running separately.

**Comment by potiuk** [comment]
> I want to suggest finding a home for that diagram in the docs, I really like that.

Big +1 (of course it should be "correct" version). As mentioned on slack I am advocating for more similar diagrams describing our code/architecture/decisions following the diagrams that can be generated automatically from sources with mermaid being at clear advantage as being supported by GitHub natively (See https://github.com/apache/airflow/blob/main/airflow/jobs/JOB_LIFECYCLE.md)

**Comment by ferruzzi** [comment]
> However, that's not "adoption" time. Adoption refers to tasks a scheduler adopts from a no-longer-running scheduler.

Sometimes it feels like naming things is the hardest part of the job.   Do you have any suggestions?

**Comment by jedcunningham** [comment]
Absolutely! I've been involved with discussions recently around caching and now naming. Makes me think of [two things](https://martinfowler.com/bliki/TwoHardThings.html).

If we split them, maybe `scheduled_duration` and `queued_duration`?

**Comment by potiuk** [comment]
> Absolutely! I've been involved with discussions recently around caching and now naming. Makes me think of [two things](https://martinfowler.com/bliki/TwoHardThings.html).

I think @jedcunningham you meant THREE things (or was it off-by-one?)

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (vandonr-amz) [reply airflow/models/taskinstance.py]: I followed what is done everywhere else in the code, one metric is emitted with the "variables" in the name (the "legacy" metric), and a similar one with a constant name and the variables in tags.

**Comment by vandonr-amz** [comment]
I'm working on splitting this in 2, and I have a question:
in case a state is skipped, for instance here:
 https://github.com/apache/airflow/blob/44034198c8d9a03dee28c8f2167c347c4f801dd2/airflow/jobs/backfill_job_runner.py#L530-L534
should we emit a metric that the task spent 0s in scheduled state ? Or should we not emit at all ?
I'm taking the "no metric" approach for now, but I'm not convinced it's the best.

File: airflow/config_templates/config.yml [hunk 2]
```diff
@@ -2052,8 +2052,9 @@ celery:
         Time in seconds after which adopted tasks which are queued in celery are assumed to be stalled,
         and are automatically rescheduled. This setting does the same thing as ``stalled_task_timeout`` but
         applies specifically to adopted tasks only. When set to 0, the ``stalled_task_timeout`` setting
-        also applies to adopted tasks. To calculate adoption time, subtract the
-        :ref:`task duration<ui:task-duration>` from the task's :ref:`landing time<ui:landing-times>`.
+        also applies to adopted tasks.
+        To calculate adoption time, add the ``dag.<dag_id>.<task_id>.scheduled_duration`` and
+        ``dag.<dag_id>.<task_id>.queued_duration`` metrics.
```
Comment (vandonr-amz) [inline airflow/config_templates/config.yml:2057]: did I get this right ?

**Comment by potiuk** [comment]
No metrics would be my choice too.

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -1387,6 +1389,52 @@ def _log_state(self, lead_msg: str = "") -> None:
             self._date_or_empty("end_date"),
         )
 
+    def emit_state_change_metric(self, new_state: TaskInstanceState):
+        """
+        Sends a time metric logging how much time a given state transition took.
```
Comment (o-nikolas) [inline airflow/models/taskinstance.py]:
```suggestion
        Sends a time metric representing how much time a given state transition took.
```

File: airflow/models/taskinstance.py [hunk 4]
```diff
@@ -1387,6 +1389,52 @@ def _log_state(self, lead_msg: str = "") -> None:
...
+                # same comment as above
+                self.log.warning(
+                    "cannot record %s for task %s because previous state change time has not been saved",
+                    metric_name,
+                    self.task_id,
+                )
+                return
+            timing = (timezone.utcnow() - self.start_date).total_seconds()
+        else:
+            raise NotImplementedError
```
Comment (o-nikolas) [inline airflow/models/taskinstance.py]: Maybe add a string to the error that provides a bit more context and info. you can add which state was provided and which states are supported.

File: airflow/models/taskinstance.py [hunk 5]
```diff
@@ -1387,6 +1389,52 @@ def _log_state(self, lead_msg: str = "") -> None:
...
+
+        # send metric twice, once (legacy) with tags in the name and once with tags as tags
+        Stats.timing(f"dag.{self.dag_id}.{self.task_id}.{metric_name}", timing)
+        Stats.timing(f"task.{metric_name}", timing, tags={"task_id": self.task_id, "dag_id": self.dag_id})
```
Comment (o-nikolas) [inline airflow/models/taskinstance.py:1441]: 👍

**Review by o-nikolas — COMMENTED** [review]
Looks good, left a few nits 👍

File: docs/apache-airflow/administration-and-deployment/logging-monitoring/metrics.rst [hunk 6]
```diff
@@ -164,7 +164,9 @@ Timers
 Name                                                Description
 =================================================== ========================================================================
 ``dagrun.dependency-check.<dag_id>``                Milliseconds taken to check DAG dependencies
-``dag.<dag_id>.<task_id>.duration``                 Seconds taken to finish a task
+``dag.<dag_id>.<task_id>.duration``                 Seconds taken to run a task
+``dag.<dag_id>.<task_id>.scheduled_duration``       Seconds a task spends in the Scheduled state, before being Queued
+``dag.<dag_id>.<task_id>.queued_duration``          Seconds a task spends in the Queued state, before being Running
```
Comment (ferruzzi) [inline docs/apache-airflow/administration-and-deployment/logging-monitoring/metrics.rst:169]: (Non-blocking) Just curious, are these lines manual changes or from some automated change?

**Review by ferruzzi — APPROVED** [review]
Left a question out of curiosity, but no blocking requests.

File: airflow/models/taskinstance.py (hunk 4 above)
Comment (vandonr-amz) [reply airflow/models/taskinstance.py:1432]: also now that I look at it, I think I was raising the type instead of an exception, it's missing the `()` 😬

Comment (vandonr-amz) [reply airflow/models/taskinstance.py]: I'll add a bit more context, but it's an error for developers, not for customers, so I don't think too much info is needed.

File: docs/apache-airflow/administration-and-deployment/logging-monitoring/metrics.rst (hunk 6 above)
Comment (vandonr-amz) [reply docs/apache-airflow/administration-and-deployment/logging-monitoring/metrics.rst:169]: I added them manually, there is nothing enforcing sync between this doc and the code.

File: airflow/models/taskinstance.py (hunk 4 above)
Comment (o-nikolas) [reply airflow/models/taskinstance.py]: As a developer I appreciate helpful tracebacks 🙂

**Review by o-nikolas — APPROVED** [review]

File: docs/apache-airflow/administration-and-deployment/logging-monitoring/metrics.rst (hunk 6 above)
Comment (ferruzzi) [reply docs/apache-airflow/administration-and-deployment/logging-monitoring/metrics.rst:169]: Alright, that was what I thought.  I went looking for something like this when I started on the OTel work and didn't find it.  That's buried pretty deep.  :+1:

**Review by potiuk — APPROVED** [review]

**Comment by htpawel** [comment]
@vandonr-amz @potiuk Hi,
Above change has two bugs. One is that Stats.timing expects milliseconds instead of seconds (PR with fix: [37936](https://github.com/apache/airflow/pull/37936/files)) and second one is that scheduled_duration metric does not work as `self.start_date` is always None at time of execution of this. It is because `self.start_date` is not corresponding to time of schedule (real start) but to actual running start (after scheduled and queued states), see comment here: [34771#discussion_r1432601056](https://github.com/apache/airflow/pull/34771#discussion_r1432601056). I think something with naming convention should be done as it can be very misleading, and also we need some scheduled_dttm variable like stated in the comment.

**Comment by ferruzzi** [comment]
@htpawel  This is quite old, would you please open [a new Issue](https://github.com/apache/airflow/issues/new/choose) with your bug report?

**Pull request merged**